### PR TITLE
feat(client, sdk, cli): env loading rewrite

### PR DESCRIPTION
**Pull request by williamluke4**

### CODE REVIEW

File: src/packages/cli/src/__tests__/__snapshots__/dotenv-1-custom-schema-path.test.ts.snap [hunk 1]
```diff
@@ -4,7 +4,7 @@ exports[`should read .env file in root folder and custom-path 1`] = `
 Array [
   Array [
     Environment variables loaded from /tmp/dir/.env
```
Comment (timsuchanek) [inline src/packages/cli/src/__tests__/__snapshots__/dotenv-1-custom-schema-path.test.ts.snap:6]: General question: Why is this an array of arrays?

File: src/packages/client/src/generation/TSClient.ts [hunk 2]
```diff
@@ -82,6 +83,7 @@ const {
 } = require('${runtimePath}')
 
 const path = require('path')
+const fs = require('fs')
```
Comment (timsuchanek) [inline src/packages/client/src/generation/TSClient.ts:86]: It looks like this is not used?
```suggestion
```

File: src/packages/client/src/generation/TSClient.ts [hunk 3]
```diff
@@ -281,9 +283,16 @@ export class TSClient implements Generatable {
...
     } = this.options
+    const schemaPath = path.join(schemaDir, 'prisma.schema')
+    const envPaths = getEnvPaths(schemaPath, {cwd: outputDir})
+    const relativeEnvPaths = {
+      rootEnvPath: envPaths.rootEnvPath && path.relative(outputDir, envPaths.rootEnvPath),
+      schemaEnvPath: envPaths.schemaEnvPath && path.relative(outputDir, envPaths.schemaEnvPath)
+    }
 
-    const config: Omit<GetPrismaClientOptions, 'document' | 'dirname'> = {
+    const config: Omit<GetPrismaClientOptions, 'document' | 'dirname' > = {
```
Comment (timsuchanek) [inline src/packages/client/src/generation/TSClient.ts:293]: Let's remove the space again :D (unless prettier added it)
```suggestion
    const config: Omit<GetPrismaClientOptions, 'document' | 'dirname'> = {
```

File: src/packages/sdk/src/utils/loadEnv.ts [hunk 4]
```diff
@@ -49,20 +42,35 @@ export function tryLoadEnv(
     './prisma/.env', // 3 - Check ./prisma directory for .env
     './.env', // 4 - Check cwd for .env
   ]
-  let schemaEnvInfo: LoadEnvResult | null = null
+  let schemaEnvPath: string | null = null
   for (const envPath of schemaEnvPaths) {
```
Comment (timsuchanek) [inline src/packages/sdk/src/utils/loadEnv.ts:46]: Using `schemaEnvPaths.find` instead would probably be more readable

**Comment by timsuchanek** [comment]
Looks good so far!
Two points:
1. In your own words, can you describe, why we need two functions now `getEnvPaths` and `tryLoadEnvs`?
2. Can you add basic tests for this? (Optimally simulating the Blitz situation)

**Comment by williamluke4** [comment]
`getEnvPath` is used so that the same logic can be used on the client and the cli . The same goes for `tryLoadEnvs`

File: src/packages/cli/src/__tests__/__snapshots__/dotenv-1-custom-schema-path.test.ts.snap (hunk 1 above)
Comment (williamluke4) [reply src/packages/cli/src/__tests__/__snapshots__/dotenv-1-custom-schema-path.test.ts.snap:6]: Because I reused logic from other tests

Comment (williamluke4) [reply src/packages/cli/src/__tests__/__snapshots__/dotenv-1-custom-schema-path.test.ts.snap]: Fixed :)

**Comment by williamluke4** [comment]
@timsuchanek I've done some thinking about the blitz test and I think that we should just add it to the e2e tests as it is highly likely that Blitz will change a lot. So I don't think that a static test would achieve very much in the long term.

**Comment by williamluke4** [comment]
closing in favour of #4142

**Pull request closed without merging**